node/src/store/indexed_db_store.rs

```diff
@@ -537,10 +572,128 @@ async fn verify_against_neighbours(
     Ok(())
 }
 
+/// Similar to `get_schema_version` but detects older schema versions.
+async fn get_schema_version_from_db(db: &Rexie) -> Result<Option<u32>> {
+    let tx = db.transaction(
+        &[HEADER_STORE_NAME, RANGES_STORE_NAME, SCHEMA_STORE_NAME],
+        TransactionMode::ReadOnly,
+    )?;
+    let schema_store = tx.store(SCHEMA_STORE_NAME)?;
+    let ranges_store = tx.store(RANGES_STORE_NAME)?;
+    let header_store = tx.store(HEADER_STORE_NAME)?;
+
+    // Schema version exists from v4 and above.
+    if let Ok(version) = get_schema_version(&schema_store).await {
+        return Ok(Some(version));
+    }
+
+    // If schema version does not exist in db but ranges store
+    // has values then assume we are in version 3.
+    if !store_is_empty(&ranges_store).await? {
+        return Ok(Some(3));
+    }
+
+    // If ranges store does not have any values but header for height 1
+    // exists, then we assume we are in version 2.
+    let height_key = to_value(&1)?;
+    let height_index = header_store.index(HEIGHT_INDEX_NAME)?;
+    if height_index.get(&height_key).await?.is_truthy() {
+        return Ok(Some(2));
+    }
+
+    // Otherwise we assume we have a new db.
+    Ok(None)
+}
+
+async fn store_is_empty(store: &rexie::Store) -> Result<bool> {
+    let vals = store.get_all(None, Some(1), None, None).await?;
+    Ok(vals.is_empty())
+}
+
+async fn get_schema_version(store: &rexie::Store) -> Result<u32> {
+    let key = to_value(VERSION_KEY)?;
+    let val = store.get(&key).await?;
+    Ok(from_value(val)?)
+}
+
+async fn set_schema_version(store: &rexie::Store, version: u32) -> Result<()> {
+    let key = to_value(VERSION_KEY)?;
+    let val = to_value(&version)?;
+    store.put(&val, Some(&key)).await?;
+    Ok(())
+}
+
+async fn get_last_header_v2(store: &rexie::Store) -> Result<ExtendedHeader> {
```

Comment:
Maybe we could organise old, version-specific `fn` in modules, like `v2::get_last_header`, `v4::migrate_to`?

Reply:
I aggre with the `get_last_header` but for the migration function I'm not really sure.